File: src/MbtiEnterpriseSimilarity.App/AppOptions.cs
using System.Globalization;
using MbtiEnterpriseSimilarity.App.Domain;

namespace MbtiEnterpriseSimilarity.App;

public sealed record AppOptions(
    string InputPath,
    string TargetId,
    int TopN,
    string OutputDirectory,
    double MaxSkippedRatio,
    SimilarityMode SimilarityMode,
    DimensionWeights Weights,
    IReadOnlyList<string> ExcludedIds,
    bool ShowHelp)
{
    public static AppOptions Parse(string[] args)
    {
        if (args.Length == 0)
        {
            return new AppOptions(string.Empty, string.Empty, 5, string.Empty, 1d, SimilarityMode.Raw, DimensionWeights.Equal, [], true);
        }

        string? inputPath = null;
        string? targetId = null;
        var topN = 5;
        string? outputDirectory = null;
        var maxSkippedRatio = 1d;
        var mode = SimilarityMode.Raw;
        var weights = DimensionWeights.Equal;
        var excludedIds = new HashSet<string>(StringComparer.OrdinalIgnoreCase);

        for (var i = 0; i < args.Length; i++)
        {
            var current = args[i];

            switch (current)
            {
                case "--help":
                case "-h":
                    return new AppOptions(string.Empty, string.Empty, 5, string.Empty, 1d, SimilarityMode.Raw, DimensionWeights.Equal, [], true);

                case "--input":
                    inputPath = ReadRequiredValue(args, ref i, "--input");
                    break;

                case "--target-id":
                    targetId = ReadRequiredValue(args, ref i, "--target-id");
                    break;

                case "--top":
                    var topRaw = ReadRequiredValue(args, ref i, "--top");
                    if (!int.TryParse(topRaw, out topN) || topN <= 0)
                    {
                        throw new ArgumentException("--top must be a positive integer.");
                    }

                    break;

                case "--output-dir":
                    outputDirectory = ReadRequiredValue(args, ref i, "--output-dir");
                    break;

                case "--max-skipped-ratio":
                    var skippedRatioRaw = ReadRequiredValue(args, ref i, "--max-skipped-ratio");
                    maxSkippedRatio = ParseMaxSkippedRatio(skippedRatioRaw);
                    break;

                case "--exclude-id":
                    var excludeRaw = ReadRequiredValue(args, ref i, "--exclude-id");
                    AddExcludedIds(excludedIds, excludeRaw);
                    break;

                case "--mode":
                    var modeRaw = ReadRequiredValue(args, ref i, "--mode");
                    mode = ParseMode(modeRaw);
                    break;

                case "--weights":
                    var weightsRaw = ReadRequiredValue(args, ref i, "--weights");
                    weights = DimensionWeights.Parse(weightsRaw);
                    break;

                default:
                    throw new ArgumentException($"Unknown argument: {current}");
            }
        }

        if (string.IsNullOrWhiteSpace(inputPath))
        {
            throw new ArgumentException("--input is required.");
        }

        if (string.IsNullOrWhiteSpace(targetId))
        {
            throw new ArgumentException("--target-id is required.");
        }

        outputDirectory ??= Path.Combine(Environment.CurrentDirectory, "output");

        return new AppOptions(
            InputPath: Path.GetFullPath(inputPath),
            TargetId: targetId.Trim(),
            TopN: topN,
            OutputDirectory: Path.GetFullPath(outputDirectory),
            MaxSkippedRatio: maxSkippedRatio,
            SimilarityMode: mode,
            Weights: weights,
            ExcludedIds: excludedIds.OrderBy(id => id, StringComparer.Ordinal).ToList(),
            ShowHelp: false);
    }

    public static void PrintUsage()
    {
        Console.WriteLine("MBTI Enterprise Similarity Analyzer");
        Console.WriteLine();
        Console.WriteLine("Usage:");
        Console.WriteLine("  dotnet run --project src/MbtiEnterpriseSimilarity.App -- \\");
        Console.WriteLine("    --input <csv-path> --target-id <student-id> [--top 5] [--output-dir <folder>] [--max-skipped-ratio 0.2] [--exclude-id <id>] [--mode raw|zscore] [--weights Ne=1,Ni=1,Te=1,Ti=1,Se=1,Si=1,Fe=1,Fi=1]");
        Console.WriteLine();
        Console.WriteLine("Examples:");
        Console.WriteLine("  dotnet run --project src/MbtiEnterpriseSimilarity.App -- \\");
        Console.WriteLine("    --input \"./data/CSS121_MBTI_2026_68.csv\" --target-id 68090500418 --top 5 --max-skipped-ratio 0.2 --exclude-id 99999999999 --mode zscore");
        Console.WriteLine("  dotnet run --project src/MbtiEnterpriseSimilarity.App -- \\");
        Console.WriteLine("    --input \"./data/CSS121_MBTI_2026_68.csv\" --target-id 68090500418 --mode zscore --weights \"Ne=1.2,Ni=1.2,Te=1,Ti=1,Se=0.8,Si=0.8,Fe=1,Fi=1\"");
    }

    private static string ReadRequiredValue(string[] args, ref int index, string flag)
    {
        if (index + 1 >= args.Length)
        {
            throw new ArgumentException($"{flag} requires a value.");
        }

        index++;
        return args[index];
    }

    private static void AddExcludedIds(ISet<string> destination, string rawValue)
    {
        foreach (var token in rawValue.Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries))
        {
            if (!string.IsNullOrWhiteSpace(token))
            {
                destination.Add(token);
            }
        }
    }

    private static SimilarityMode ParseMode(string rawValue)
    {
        if (rawValue.Equals("raw", StringComparison.OrdinalIgnoreCase))
        {
            return SimilarityMode.Raw;
        }

        if (rawValue.Equals("zscore", StringComparison.OrdinalIgnoreCase) ||
            rawValue.Equals("z-score", StringComparison.OrdinalIgnoreCase))
        {
            return SimilarityMode.ZScore;
        }

        throw new ArgumentException("--mode must be either 'raw' or 'zscore'.");
    }

    private static double ParseMaxSkippedRatio(string rawValue)
    {
        if (!double.TryParse(rawValue, NumberStyles.Float, CultureInfo.InvariantCulture, out var ratio))
        {
            throw new ArgumentException("--max-skipped-ratio must be a number between 0 and 1.");
        }

        if (double.IsNaN(ratio) || double.IsInfinity(ratio) || ratio < 0 || ratio > 1)
        {
            throw new ArgumentException("--max-skipped-ratio must be in range [0, 1].");
        }

        return ratio;
    }
}

File: src/main/java/com/mbti/similarity/app/AppOptions.java
package com.mbti.similarity.app;

import com.mbti.similarity.domain.DimensionWeights;
import com.mbti.similarity.domain.SimilarityMode;
import java.io.PrintStream;
import java.nio.file.Path;
import java.util.Collection;
import java.util.Comparator;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Objects;

public final class AppOptions {
    private final Path inputPath;
    private final String targetId;
    private final int topN;
    private final Path outputDirectory;
    private final double maxSkippedRatio;
    private final SimilarityMode similarityMode;
    private final DimensionWeights weights;
    private final List<String> excludedIds;
    private final boolean showHelp;

    private AppOptions(
        Path inputPath,
        String targetId,
        int topN,
        Path outputDirectory,
        double maxSkippedRatio,
        SimilarityMode similarityMode,
        DimensionWeights weights,
        List<String> excludedIds,
        boolean showHelp
    ) {
        this.inputPath = inputPath;
        this.targetId = targetId;
        this.topN = topN;
        this.outputDirectory = outputDirectory;
        this.maxSkippedRatio = maxSkippedRatio;
        this.similarityMode = similarityMode;
        this.weights = weights;
        this.excludedIds = List.copyOf(excludedIds);
        this.showHelp = showHelp;
    }

    public static AppOptions parse(String[] args) {
        if (args == null || args.length == 0) {
            return help();
        }

        String inputPath = null;
        String targetId = null;
        int topN = 5;
        String outputDirectory = null;
        double maxSkippedRatio = 1d;
        SimilarityMode mode = SimilarityMode.RAW;
        DimensionWeights weights = DimensionWeights.EQUAL;
        LinkedHashSet<String> excludedIds = new LinkedHashSet<>();

        for (int index = 0; index < args.length; index++) {
            String current = args[index];
            switch (current) {
                case "--help":
                case "-h":
                    return help();
                case "--input":
                    inputPath = readRequiredValue(args, ++index, "--input");
                    break;
                case "--target-id":
                    targetId = readRequiredValue(args, ++index, "--target-id");
                    break;
                case "--top":
                    topN = parseTopN(readRequiredValue(args, ++index, "--top"));
                    break;
                case "--output-dir":
                    outputDirectory = readRequiredValue(args, ++index, "--output-dir");
                    break;
                case "--max-skipped-ratio":
                    maxSkippedRatio = parseMaxSkippedRatio(readRequiredValue(args, ++index, "--max-skipped-ratio"));
                    break;
                case "--exclude-id":
                    addExcludedIds(excludedIds, readRequiredValue(args, ++index, "--exclude-id"));
                    break;
                case "--mode":
                    mode = SimilarityMode.parse(readRequiredValue(args, ++index, "--mode"));
                    break;
                case "--weights":
                    weights = DimensionWeights.parse(readRequiredValue(args, ++index, "--weights"));
                    break;
                default:
                    throw new IllegalArgumentException("Unknown argument: " + current);
            }
        }

        if (inputPath == null || inputPath.isBlank()) {
            throw new IllegalArgumentException("--input is required.");
        }

        if (targetId == null || targetId.isBlank()) {
            throw new IllegalArgumentException("--target-id is required.");
        }

        Path normalizedOutputDirectory = outputDirectory == null || outputDirectory.isBlank()
            ? Path.of(System.getProperty("user.dir"), "output")
            : Path.of(outputDirectory).toAbsolutePath().normalize();

        List<String> normalizedExcludedIds = excludedIds.stream()
            .filter(Objects::nonNull)
            .map(String::trim)
            .filter(value -> !value.isEmpty())
            .sorted(Comparator.naturalOrder())
            .toList();

        return new AppOptions(
            Path.of(inputPath).toAbsolutePath().normalize(),
            targetId.trim(),
            topN,
            normalizedOutputDirectory,
            maxSkippedRatio,
            mode,
            weights,
            normalizedExcludedIds,
            false
        );
    }

    public static void printUsage(PrintStream output) {
        output.println("MBTI Similarity Analyzer");
        output.println();
        output.println("Usage:");
        output.println("  java -cp build/classes/main com.mbti.similarity.Main \\");
        output.println(
            "    --input <csv-path> --target-id <student-id> [--top 5] [--output-dir <folder>] "
                + "[--max-skipped-ratio 0.2] [--exclude-id <id>] [--mode raw|zscore] "
                + "[--weights Ne=1,Ni=1,Te=1,Ti=1,Se=1,Si=1,Fe=1,Fi=1]"
        );
        output.println();
        output.println("Examples:");
        output.println("  java -cp build/classes/main com.mbti.similarity.Main \\");
        output.println(
            "    --input \"./data/CSS121_MBTI_2026_68_2.csv\" --target-id 68090500418 --top 5 "
                + "--max-skipped-ratio 0.2 --exclude-id 99999999999 --mode zscore"
        );
        output.println("  java -cp build/classes/main com.mbti.similarity.Main \\");
        output.println(
            "    --input \"./data/CSS121_MBTI_2026_68_2.csv\" --target-id 68090500418 --mode zscore "
                + "--weights \"Ne=1.2,Ni=1.2,Te=1,Ti=1,Se=0.8,Si=0.8,Fe=1,Fi=1\""
        );
    }

    public Path getInputPath() {
        return inputPath;
    }

    public String getTargetId() {
        return targetId;
    }

    public int getTopN() {
        return topN;
    }

    public Path getOutputDirectory() {
        return outputDirectory;
    }

    public double getMaxSkippedRatio() {
        return maxSkippedRatio;
    }

    public SimilarityMode getSimilarityMode() {
        return similarityMode;
    }

    public DimensionWeights getWeights() {
        return weights;
    }

    public List<String> getExcludedIds() {
        return excludedIds;
    }

    public boolean isShowHelp() {
        return showHelp;
    }

    private static AppOptions help() {
        return new AppOptions(
            Path.of(""),
            "",
            5,
            Path.of(""),
            1d,
            SimilarityMode.RAW,
            DimensionWeights.EQUAL,
            List.of(),
            true
        );
    }

    private static String readRequiredValue(String[] args, int index, String flag) {
        if (index >= args.length) {
            throw new IllegalArgumentException(flag + " requires a value.");
        }
        return args[index];
    }

    private static int parseTopN(String rawValue) {
        try {
            int topN = Integer.parseInt(rawValue);
            if (topN <= 0) {
                throw new IllegalArgumentException("--top must be a positive integer.");
            }
            return topN;
        } catch (NumberFormatException ex) {
            throw new IllegalArgumentException("--top must be a positive integer.", ex);
        }
    }

    private static double parseMaxSkippedRatio(String rawValue) {
        try {
            double ratio = Double.parseDouble(rawValue);
            if (!Double.isFinite(ratio) || ratio < 0d || ratio > 1d) {
                throw new IllegalArgumentException("--max-skipped-ratio must be in range [0, 1].");
            }
            return ratio;
        } catch (NumberFormatException ex) {
            throw new IllegalArgumentException("--max-skipped-ratio must be a number between 0 and 1.", ex);
        }
    }

    private static void addExcludedIds(Collection<String> destination, String rawValue) {
        for (String token : rawValue.split(",")) {
            String normalized = token.trim();
            if (!normalized.isEmpty()) {
                destination.add(normalized);
            }
        }
    }
}
